### ISSUE TITLE
Support for custom models

### ISSUE THREAD

**Issue by mlukasik-dev** [kyleconroy/sqlc#755]
Example:

`db/queries/roles.sql`
```sql
-- name: ListRolesWithPermissions :many
-- model: Roles
-- path: internal/domain/roles
SELECT
  r.role_id,
  r.display_name,
  r.description,
  array_agg(permissions.method)::text[] AS permissions
  FROM roles r
    LEFT JOIN permissions USING (role_id)
GROUP BY r.role_id;

-- name ListRoles :many
-- model: Roles
-- path: internal/domain/roles
SELECT  role_id, display_name, description FROM roles; /* returned colums exists in the model so we can reuse it */
```

It isn't generated file!
`internal/domain/roles`
```go
type Role struct {
	RoleID      uuid.UUID // here we can add some custom struct tags and methods
	DisplayName string
	Description string
	Permissions []sql.NullString // if array can is empty using []string throws error
}
```

Generated method
```go
func (q *Queries) ListRoles(ctx context.Context) ([]roles.Role, error)
```

P.S. This features can be useful because with it we can both use sqlc and follow **Clean Architecture**. And reuse structs for tables for complex queries with joins.

So my question is if I'd correctly implement this functionality would you approve PR (I'm asking because this feature can be not in the spirit of the project)

**Comment by mlukasik-dev** [comment]
This feature could help as solve issues like: #185 , #534 , #658

**Comment by jonleopard** [comment]
Would also love this! I use sqlx and frequently use tags like   `db:"user_id" json:"id"` or `json:"password_confirm" validate:"eqfield=Password"`

**Comment by Jille** [comment]
I agree this would be useful. Especially when you have a few queries with different WHERE clauses that return the same columns.

This proposal has one big question: How would we verify that the given struct matches the data types returned from the query?
* Do we assert they match exactly with the types sqlc would've chosen? That'd be safe, but I think a bunch of people here are expecting to use this to override exactly that? Also, if the types match 1:1, Go would allow you to cast it anyway.
* If we allow uncheck structs, wouldn't that defeat one of sqlc's USPs: That you don't have runtime failures for type mismatches.

Would love to hear your thoughts on this.

**Comment by jamesleeht** [comment]
Also very interested in this feature. Here's another related issue: #444
